File: app/src/main/java/com/mx/gillustrated/database/imp/CardInfoDaoImp.java
package com.mx.gillustrated.database.imp;

import android.content.Context;
import com.j256.ormlite.dao.Dao;
import com.j256.ormlite.dao.RawRowMapper;
import com.j256.ormlite.dao.RuntimeExceptionDao;
import com.j256.ormlite.stmt.DeleteBuilder;
import com.j256.ormlite.stmt.QueryBuilder;
import com.j256.ormlite.stmt.UpdateBuilder;
import com.j256.ormlite.stmt.Where;
import com.mx.gillustrated.activity.BaseActivity;
import com.mx.gillustrated.database.DataBaseHelper;
import com.mx.gillustrated.vo.CardEventInfo;
import com.mx.gillustrated.vo.CardInfo;

import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by maoxin on 2018/7/24.
 */

public class CardInfoDaoImp  extends RuntimeExceptionDao<CardInfo, Integer> {

    private static Dao<CardInfo, Integer> mDao = null;
    private DataBaseHelper mOrm = null;
    private Context mContext;

    private static Dao<CardInfo, Integer> getDao(DataBaseHelper orm) {
        if (mDao == null){
            try {
                mDao = orm.getDao(CardInfo.class);
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
        return mDao;
    }

    public CardInfoDaoImp(DataBaseHelper orm) {
        super(CardInfoDaoImp.getDao(orm));
        this.mOrm = orm;
        this.mContext = orm.mContext;
    }

    public void addCardInfos(List<CardInfo> infos){
        for(CardInfo info: infos){
            this.createIfNotExists(info);
        }
    }

    public void delCardsById(List<Integer> list){
        for(int id:list){
            this.deleteById(id);
        }
    }

    public int delCardInfoByGameId(int gameId){
        DeleteBuilder<CardInfo, Integer> db = this.deleteBuilder();
        Where<CardInfo, Integer> where = db.where();
        try {
            where.eq(CardInfo.COLUMN_GAMETYPE, gameId);
            return this.delete(db.prepare());
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return -1;
    }

    public List<CardInfo> queryCards(CardInfo cardinfo, String orderBy, boolean isAsc, long position, long limit) {
        List<CardInfo> result = null;
        QueryBuilder<CardInfo, Integer> qb = this.queryBuilder();
        boolean showHeader = mContext.getSharedPreferences("commonset", Context.MODE_PRIVATE).getBoolean(BaseActivity.SHARE_SHOW_HEADER_IMAGES, false);
        if(orderBy.equals(CardInfo.COLUMN_NID) && !showHeader)
            orderBy = CardInfo.COLUMN_TOTAL;
        //qb.orderBy(orderBy, isAsc);
        qb.orderByRaw(orderBy + " " + ( isAsc ? "ASC" : "DESC") );
        //qb.selectRaw("*");
        qb.selectRaw("*," + CardInfo.COLUMN_MAXATTACK + "+" +  CardInfo.COLUMN_MAXDEFENSE + "+" + CardInfo.COLUMN_MAXHP + "+" + CardInfo.COLUMN_EXTRA_VALUE1
                + "+" + CardInfo.COLUMN_EXTRA_VALUE2 + " " + CardInfo.COLUMN_TOTAL);
//        qb.selectRaw("(SELECT " + CardTypeInfo.COLUMN_NAME + " FROM " + CardTypeInfo.TABLE_NAME + " where " +
//                CardTypeInfo.TABLE_NAME +  "." + CardTypeInfo.ID + " = " + CardInfo.TABLE_NAME + "." + CardInfo.COLUMN_ATTR + ") ATTRNAME" );
        try {

            qb.limit(limit);
            qb.offset(position);

            Where<CardInfo, Integer> where =  qb.where();
            where.eq(CardInfo.COLUMN_GAMETYPE, cardinfo.getGameId());
            if(cardinfo.getPinyinName() != null)
                where.and().like(CardInfo.COLUMN_PINYIN_NAME, cardinfo.getPinyinName() + "%");
            if(cardinfo.getName() != null)
                where.and().like(CardInfo.COLUMN_NAME, "%" + cardinfo.getName() + "%");
            if(cardinfo.getFrontName() != null)
                where.and().eq(CardInfo.COLUMN_FRONT_NAME, cardinfo.getFrontName() );
            if(cardinfo.getId() > 0)
                where.and().eq(CardInfo.ID, cardinfo.getId());
            if(cardinfo.getNid() > 0)
                where.and().eq(CardInfo.COLUMN_NID, cardinfo.getNid());
            if(cardinfo.getAttrId() > 0)
                where.and().eq(CardInfo.COLUMN_ATTR, cardinfo.getAttrId());
            if(cardinfo.getLevel() != null)
                where.and().eq(CardInfo.COLUMN_LEVEL, cardinfo.getLevel());
            if(cardinfo.getCost() > -1)
                where.and().eq(CardInfo.COLUMN_COST, cardinfo.getCost());
            if(cardinfo.getEventId() > 0){
                where.and().raw("EXISTS ( SELECT * FROM  "
                        + CardEventInfo.TABLE_NAME
                        + " WHERE " + CardInfo.TABLE_NAME + "." + CardInfo.ID + " = "
                        + CardEventInfo.TABLE_NAME + "." + CardEventInfo.COLUMN_CARD_NID + " and "
                        + CardEventInfo.TABLE_NAME + "." + CardEventInfo.COLUMN_EVENT_ID + " = " + cardinfo.getEventId() + "  ) ");
            }
            result = this.queryRaw(qb.prepareStatementString(), new CardInfoRowMapper()).getResults();

            //计算总数
            QueryBuilder<CardInfo, Integer> qbCount = this.queryBuilder();
            qbCount.setCountOf(true);
            qbCount.selectRaw("*");
            qbCount.setWhere(where);
            int count = (int) this.countOf(qbCount.prepare());

            for( CardInfo cardInfo : result ){
                cardInfo.setAttr(getAttrName(cardInfo.getAttrId()));
                cardInfo.setTotalCount(count);
            }

        } catch (SQLException e) {
            e.printStackTrace();
        }
        return result;
    }

    private String getAttrName(int id){
        return this.mOrm.getCardTypeInfoDao().queryForId(id).getName();
    }

    public List<CardInfo> queryCardDropList(String type, int gametype) {
        List<String[]> result = new ArrayList<>();
        QueryBuilder<CardInfo, Integer> qb = this.queryBuilder();
        qb.selectRaw("DISTINCT ( " + type + ") ");
        qb.selectRaw("COUNT (*)");
        qb.groupBy(type);
        try {
            qb.where().eq(CardInfo.COLUMN_GAMETYPE, gametype);
            result = qb.queryRaw().getResults();
        } catch (SQLException e) {
            e.printStackTrace();
        }

        List<CardInfo> cardInfos = new ArrayList();
        for( String[] row : result ){
            CardInfo cardInfo = new CardInfo();
            if(type == CardInfo.COLUMN_ATTR ) {
                cardInfo.setName(getAttrName(Integer.parseInt(row[0])));
                cardInfo.setAttrId(Integer.parseInt(row[0]));
            }
            else
                cardInfo.setName(row[0]);
            cardInfo.setNid(Integer.parseInt(row[1]));
            cardInfos.add(cardInfo);
        }
        return cardInfos;
    }

    public long updateCardName(CardInfo cardinfoNew, CardInfo cardinfoOld)
    {
        UpdateBuilder<CardInfo, Integer> ub = this.updateBuilder();
        try {
            ub.updateColumnValue(CardInfo.COLUMN_NAME, cardinfoNew.getName());
            ub.updateColumnValue(CardInfo.COLUMN_PINYIN_NAME, cardinfoNew.getPinyinName());
            ub.where().eq(CardInfo.COLUMN_NAME, cardinfoOld.getName());
            return this.update(ub.prepare());
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return -1;
    }


    private static class CardInfoRowMapper implements RawRowMapper<CardInfo>{

        @Override
        public CardInfo mapRow(String[] columnNames, String[] resultColumns) throws SQLException {
            CardInfo info = new CardInfo();
            for(int i = 0; i < columnNames.length; i++){
                switch (columnNames[i]){
                    case CardInfo.ID:
                        info.setId(Integer.parseInt(resultColumns[i]));
                        break;
                    case CardInfo.COLUMN_ATTR:
                        info.setAttrId(Integer.parseInt(resultColumns[i]));
                        break;
                    case CardInfo.COLUMN_COST:
                        info.setCost(Integer.parseInt(resultColumns[i]));
                        break;
                    case CardInfo.COLUMN_FRONT_NAME:
                        info.setFrontName(resultColumns[i]);
                        break;
                    case CardInfo.COLUMN_GAMETYPE:
                        info.setGameId(Integer.parseInt(resultColumns[i]));
                        break;
                    case CardInfo.COLUMN_LEVEL:
                        info.setLevel(resultColumns[i]);
                        break;
                    case CardInfo.COLUMN_MAXATTACK:
                        info.setMaxAttack(resultColumns[i]);
                        break;
                    case CardInfo.COLUMN_MAXDEFENSE:
                        info.setMaxDefense(resultColumns[i]);
                        break;
                    case CardInfo.COLUMN_MAXHP:
                        info.setMaxHP(resultColumns[i]);
                        break;
                    case CardInfo.COLUMN_NAME:
                        info.setName(resultColumns[i]);
                        break;
                    case CardInfo.COLUMN_NID:
                        info.setNid(Integer.parseInt(resultColumns[i]));
                        break;
                    case CardInfo.COLUMN_PINYIN_NAME:
                        info.setPinyinName(resultColumns[i]);
                        break;
                    case CardInfo.COLUMN_REMARK:
                        info.setRemark(resultColumns[i]);
                        break;
                    case CardInfo.COLUMN_SHOW_HEAD:
                        info.setProfile(resultColumns[i]);
                        break;
                    case CardInfo.COLUMN_EXTRA_VALUE1:
                        info.setExtraValue1(resultColumns[i]);
                        break;
                    case CardInfo.COLUMN_EXTRA_VALUE2:
                        info.setExtraValue2(resultColumns[i]);
                        break;
                }

            }
            return info;
        }
    }

}

File: app/src/main/java/com/mx/gillustrated/database/imp/CardInfoDaoImp.kt
package com.mx.gillustrated.database.imp

import android.content.Context
import com.j256.ormlite.dao.Dao
import com.j256.ormlite.dao.RawRowMapper
import com.j256.ormlite.dao.RuntimeExceptionDao
import com.mx.gillustrated.activity.BaseActivity
import com.mx.gillustrated.database.DataBaseHelper
import com.mx.gillustrated.vo.CardEventInfo
import com.mx.gillustrated.vo.CardInfo
import com.mx.gillustrated.vo.CardTypeInfo
import java.sql.SQLException
import java.util.ArrayList

/**
 * Created by maoxin on 2018/7/24.
 */

class CardInfoDaoImp(orm: DataBaseHelper) : RuntimeExceptionDao<CardInfo, Int>(getDao(orm)) {
    private val mOrm: DataBaseHelper = orm
    private val mContext: Context = orm.mContext


    fun addCardInfos(infos: List<CardInfo>) {
        for (info in infos) {
            this.createIfNotExists(info)
        }
    }

    fun delCardsById(list: List<Int>) {
        for (id in list) {
            this.deleteById(id)
        }
    }

    fun delCardInfoByGameId(gameId: Int): Int {
        val db = this.deleteBuilder()
        val where = db.where()
        try {
            where.eq(CardInfo.COLUMN_GAMETYPE, gameId)
            return this.delete(db.prepare())
        } catch (e: SQLException) {
            e.printStackTrace()
        }

        return -1
    }

    fun queryCards(cardinfo: CardInfo, orderByParam: String, isAsc: Boolean, position: Long, limit: Long): List<CardInfo>? {
        var orderBy = orderByParam
        var result: List<CardInfo>? = null
        val qb = this.queryBuilder()
        val showHeader = mContext.getSharedPreferences("commonset", Context.MODE_PRIVATE).getBoolean(BaseActivity.SHARE_SHOW_HEADER_IMAGES + cardinfo.gameId, false)
        if (orderBy == CardInfo.COLUMN_NID && !showHeader)
            orderBy = CardInfo.COLUMN_TOTAL
        qb.orderByRaw(orderBy + " " + if (isAsc) "ASC" else "DESC")
        qb.selectRaw("*," + CardInfo.COLUMN_MAXATTACK + "+" + CardInfo.COLUMN_MAXDEFENSE + "+" + CardInfo.COLUMN_MAXHP + "+" + CardInfo.COLUMN_EXTRA_VALUE1
                + "+" + CardInfo.COLUMN_EXTRA_VALUE2 + " " + CardInfo.COLUMN_TOTAL + ","
                + "( SELECT " + CardTypeInfo.COLUMN_NAME  + " FROM "  + CardTypeInfo.TABLE_NAME
                + " WHERE " +  CardTypeInfo.TABLE_NAME + "." + CardTypeInfo.ID + " = " + CardInfo.TABLE_NAME + "." + CardInfo.COLUMN_ATTR
                +  " ) " + CardInfo.COLUMN_ATTR_NAME )
        try {
            qb.limit(limit)
            qb.offset(position)

            val where = qb.where()
            where.eq(CardInfo.COLUMN_GAMETYPE, cardinfo.gameId)
            if (cardinfo.pinyinName != null)
                where.and().like(CardInfo.COLUMN_PINYIN_NAME, cardinfo.pinyinName!! + "%")
            if (cardinfo.name != null)
                where.and().like(CardInfo.COLUMN_NAME, "%" + cardinfo.name + "%")
            if (cardinfo.frontName != null)
                where.and().eq(CardInfo.COLUMN_FRONT_NAME, cardinfo.frontName)
            if (cardinfo.id > 0)
                where.and().eq(CardInfo.ID, cardinfo.id)
            if (cardinfo.nid > 0)
                where.and().eq(CardInfo.COLUMN_NID, cardinfo.nid)
            if (cardinfo.attrId > 0)
                where.and().eq(CardInfo.COLUMN_ATTR, cardinfo.attrId)
            if (cardinfo.level != null)
                where.and().eq(CardInfo.COLUMN_LEVEL, cardinfo.level)
            if (cardinfo.cost > -1)
                where.and().eq(CardInfo.COLUMN_COST, cardinfo.cost)
            if (cardinfo.eventId > 0) {
                where.and().raw("EXISTS ( SELECT * FROM  "
                        + CardEventInfo.TABLE_NAME
                        + " WHERE " + CardInfo.TABLE_NAME + "." + CardInfo.ID + " = "
                        + CardEventInfo.TABLE_NAME + "." + CardEventInfo.COLUMN_CARD_NID + " and "
                        + CardEventInfo.TABLE_NAME + "." + CardEventInfo.COLUMN_EVENT_ID + " = " + cardinfo.eventId + "  ) ")
            }
            result = this.queryRaw(qb.prepareStatementString(), CardInfoRowMapper()).results

            //计算总数
            val qbCount = this.queryBuilder()
            qbCount.setCountOf(true)
            qbCount.selectRaw("*")
            qbCount.setWhere(where)
            val count = this.countOf(qbCount.prepare()).toInt()

            // used as tiers
            if(orderBy == CardInfo.COLUMN_NID && showHeader && !isAsc){
                var index = 1
                result.forEach {
                    it.totalCount = count
                    if(it.nid > 0 )
                        it.nid = index++
                }
            }else{
                result.forEach {
                    it.totalCount = count
                    it.nid = 0
                }
            }


        } catch (e: SQLException) {
            e.printStackTrace()
        }

        return result
    }

    private fun getAttrName(id: Int): String? {
        val result = this.mOrm.cardTypeInfoDao.queryForId(id)
        return if (result == null) "" else result.name
    }

    fun queryCardDropList(type: String, gametype: Int): MutableList<CardInfo> {
        var result: List<Array<String>> = ArrayList()
        val qb = this.queryBuilder()
        qb.selectRaw("DISTINCT ( $type) ")
        qb.selectRaw("COUNT (*)")
        qb.groupBy(type)
        try {
            qb.where().eq(CardInfo.COLUMN_GAMETYPE, gametype)
            result = qb.queryRaw().results
        } catch (e: SQLException) {
            e.printStackTrace()
        }

        val cardInfos = mutableListOf<CardInfo>()
        for (row in result) {
            val cardInfo = CardInfo()
            if (type === CardInfo.COLUMN_ATTR) {
                cardInfo.name = getAttrName(Integer.parseInt(row[0]))
                cardInfo.attrId = Integer.parseInt(row[0])
            } else
                cardInfo.name = row[0]
            cardInfo.nid = Integer.parseInt(row[1])
            cardInfos.add(cardInfo)
        }
        return cardInfos
    }

    fun updateCardName(cardinfoNew: CardInfo, cardinfoOld: CardInfo): Long {
        val ub = this.updateBuilder()
        try {
            ub.updateColumnValue(CardInfo.COLUMN_NAME, cardinfoNew.name)
            ub.updateColumnValue(CardInfo.COLUMN_PINYIN_NAME, cardinfoNew.pinyinName)
            ub.where().eq(CardInfo.COLUMN_NAME, cardinfoOld.name)
            return this.update(ub.prepare()).toLong()
        } catch (e: SQLException) {
            e.printStackTrace()
        }

        return -1
    }


    private class CardInfoRowMapper : RawRowMapper<CardInfo> {

        @Throws(SQLException::class)
        override fun mapRow(columnNames: Array<String>, resultColumns: Array<String>): CardInfo {
            val info = CardInfo()
            for (i in columnNames.indices) {
                when (columnNames[i]) {
                    CardInfo.ID -> info.id = Integer.parseInt(resultColumns[i])
                    CardInfo.COLUMN_ATTR -> info.attrId = Integer.parseInt(resultColumns[i])
                    CardInfo.COLUMN_COST -> info.cost = Integer.parseInt(resultColumns[i])
                    CardInfo.COLUMN_FRONT_NAME -> info.frontName = resultColumns[i]
                    CardInfo.COLUMN_GAMETYPE -> info.gameId = Integer.parseInt(resultColumns[i])
                    CardInfo.COLUMN_LEVEL -> info.level = resultColumns[i]
                    CardInfo.COLUMN_MAXATTACK -> info.maxAttack = resultColumns[i]
                    CardInfo.COLUMN_MAXDEFENSE -> info.maxDefense = resultColumns[i]
                    CardInfo.COLUMN_MAXHP -> info.maxHP = resultColumns[i]
                    CardInfo.COLUMN_NAME -> info.name = resultColumns[i]
                    CardInfo.COLUMN_NID -> info.nid = Integer.parseInt(resultColumns[i])
                    CardInfo.COLUMN_PINYIN_NAME -> info.pinyinName = resultColumns[i]
                    CardInfo.COLUMN_REMARK -> info.remark = resultColumns[i]
                    CardInfo.COLUMN_SHOW_HEAD -> info.profile = resultColumns[i]
                    CardInfo.COLUMN_EXTRA_VALUE1 -> info.extraValue1 = resultColumns[i]
                    CardInfo.COLUMN_EXTRA_VALUE2 -> info.extraValue2 = resultColumns[i]
                    CardInfo.COLUMN_ATTR_NAME  -> info.attr = resultColumns[i]
                }

            }
            return info
        }
    }

    companion object {

        private var mDao: Dao<CardInfo, Int>? = null

        private fun getDao(orm: DataBaseHelper): Dao<CardInfo, Int>? {
            if (mDao == null) {
                try {
                    mDao = orm.getDao(CardInfo::class.java)
                } catch (e: SQLException) {
                    e.printStackTrace()
                }

            }
            return mDao
        }
    }

}
